### ISSUE TITLE
System.AccessViolationException: Attempted to read or write protected memory

### ISSUE THREAD

**Issue by Lakasabasz** [myd7349/SharpLSL#1]
## Workaround:
Clone [native lib repo](https://github.com/sccn/liblsl/tree/master). Build it using your compiler, copy liblsl.dll and replace lsl.dll in bin/<Debug|Release>/net<version>/runtimes/<your platform>. Then it will work, but remember to disable native package form csproj, because it will overwrite your package with that from nuget.

---

Hi, I am trying to use modified sample from https://github.com/labstreaminglayer/liblsl-Csharp/blob/master/examples/ReceiveData/ReceiveData.cs`. My code is:
```csharp
using SharpLSL;

StreamInfo[] results = LSL.Resolve("type", "EEG");

// open an inlet and print some interesting info about the stream (meta-data, etc.)
using StreamInlet inlet = new StreamInlet(results[0]);
Console.Write(inlet.GetStreamInfo().ToXML());

// read samples
float[] sample = new float[8];
while (!Console.KeyAvailable)
{
    inlet.PullSample(sample);
    foreach (float f in sample)
        Console.Write("\t{0}", f);
    Console.WriteLine();
}
```
And stacktrace:
```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at SharpLSL.Interop.LSL.lsl_resolve_byprop(IntPtr[], UInt32, System.String, System.String, Int32, Double)
--------------------------------
   at SharpLSL.LSL.Resolve(System.String, System.String, Int32, Int32, Double)
   at Program.<Main>$(System.String[])
```

Perhaps you not using unsafe structure and garbage collector changed mount position, but not updated IntPtr.

**Comment by myd7349** [comment]
Hi! @Lakasabasz Thank you for your feedback! I’ll investigate this issue as soon as possible.

**Comment by myd7349** [comment]
Hello! @Lakasabasz After some exploration, I have to admit that I haven't pinpointed the root cause of the issue yet. Let me share my understanding for discussion (please correct me if I'm mistaken).

From what I understand, System.IntPtr is a blittable type, and its one-dimensional array is also blittable[^1]. Therefore, when using IntPtr[] as a PInvoke parameter, this argument should be automatically pinned and shouldn't be affected by the GC before the function call returns(unless `lsl_resolve_byprop` stores this address and continues to use it after the function returns)[^2]. I've tested this on both Windows 11 and WSL and couldn't reproduce the issue.

Could you please share more details?

- OS: Windows/Linux/macOS
- IDE: Visual Studio/other
- .NET Framework version/.NET version
- Project Configuration: AnyCPU/x86/x64
- SharpLSL.Native.* package version
- How is `SendData` look like?

Is the likelihood of reproducing this issue high? If so, then turn on `Enable native code debugging`[^3](assuming you are using Visual Studio) might help us identify the issue more quickly.

Additionally,SharpLSL.Native.* 1.16.2.2 packages include the PDB files corresponding to the DLLs, which were [compiled](https://github.com/myd7349/liblsl-ci-build) from [liblsl v1.16.2](https://github.com/sccn/liblsl/tree/v1.16.2) source.

Also, liblsl fixed a memory-related bug several months ago, as noted in https://github.com/sccn/liblsl/pull/211. However, this fix is not included in the v1.16.2 release. I'm not sure if this issue can be resolved with that patch.

[^1]: [Blittable and Non-Blittable Types](https://learn.microsoft.com/en-us/dotnet/framework/interop/blittable-and-non-blittable-types)
[^2]: [Pin data for P/Invoke access](https://stackoverflow.com/questions/63776096/pin-data-for-p-invoke-access)
[^3]: [Attempted to read or write protected memory. This is often an indication that other memory is corrupt](https://stackoverflow.com/questions/4074585/attempted-to-read-or-write-protected-memory-this-is-often-an-indication-that-ot)

**Comment by Lakasabasz** [comment]
Yes, I should mention about my setup. To begin with:
- Windows 11
- Rider (it's free for non-commercial use for about week)
- .NET 8.0

*.csproj:
```xml
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net8.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
    </PropertyGroup>

    <ItemGroup>
      <PackageReference Include="SharpLSL" Version="0.0.8" />
      <PackageReference Include="SharpLSL.Native.all" Version="1.16.2.2" />
    </ItemGroup>

</Project>
```
And there are no sender. It crashes on `LSL.Resolve` function.
That's should be it for debug details. I can provide you directory tree, but it look like somehow C# or native dll tries to access memory not allocated. Last time I didn't start debugger, but now I see you are using `unsafe partial class` so it suppress errors with pointer operations. 

I have 100% percent to reproduce it. So there are have to be some difference between my and your setup. I'll try to prepare data sender.

**Comment by Lakasabasz** [comment]
I was able to run it on ubuntu 22.04 so it is related to windows. I tried to run that app outside IDE and error persist. There must be something in windows native build. Let me know if you can reproduce that exception. Probably you have something installed that fixes it. Or maybe it is related to different build version of packages between your setup and nuget

**Comment by myd7349** [comment]
Hello! @Lakasabasz 

I installed Rider today and tested the code on Windows 11, but I couldn't reproduce this issue. I also tested the generated executable on a few other computers(Windows 11, some of them have Visual Studio 2022 installed, some are not), and the problem did not occur there either.

Could you check the Windows Event Viewer application logs for more information related to the crash module (e.g., module name, memory address offset). You can follow this tutorial for guidance: [Event Viewer Tutorial](https://bentleysystems.service-now.com/community?id=kb_article&sysparm_article=KB0104868).

I also suspect this might be related to the recent commit of https://github.com/myd7349/liblsl-ci-build/commit/0621013504cd9dfa2b1375af63d1dddd4ea925c1. Could you please perform the following tests?

1. Replace the `bin\Debug\runtime\win-x86\liblsl.dll` and `bin\Debug\runtime\win-x64\liblsl.dll` with the liblsl.dll from the official liblsl release:

   - https://github.com/sccn/liblsl/releases/download/v1.16.2/lsltests-1.16.2-Win_amd64.zip

   - https://github.com/sccn/liblsl/releases/download/v1.16.2/lslexamples-1.16.2-Win_i386.zip

2. Use SharpLSL.Native.* 1.16.2.1

If this still crahshes, then I suspect this might be related to https://github.com/sccn/liblsl/pull/211. Then we can compile liblsl with this patch included.

Thank you!

**Comment by myd7349** [comment]
Here are the 32-bit and 64-bit dynamic libraries generated using the latest version of liblsl.

https://github.com/myd7349/liblsl-ci-build/actions/runs/11592351741

- [liblsl-7e61a2e-win-x64.zip](https://github.com/user-attachments/files/17571525/liblsl-7e61a2e-win-x64.zip)
- [liblsl-7e61a2e-win-x86.zip](https://github.com/user-attachments/files/17571528/liblsl-7e61a2e-win-x86.zip)

**Comment by myd7349** [comment]
```csharp
using SharpLSL;

Console.WriteLine($"64 bit? {Environment.Is64BitProcess}");

// If this line of code runs successfully, it means
// the liblsl dynamic library has been loaded successfully.
Console.WriteLine($"liblsl version: {LSL.GetLibraryVersion()}");

StreamInfo[] results = LSL.Resolve("type", "EEG");

// open an inlet and print some interesting info about the stream (meta-data, etc.)
using StreamInlet inlet = new StreamInlet(results[0]);
Console.Write(inlet.GetStreamInfo().ToXML());

// read samples
float[] sample = new float[8];
while (!Console.KeyAvailable)
{
    inlet.PullSample(sample);
    foreach (float f in sample)
        Console.Write("\t{0}", f);
    Console.WriteLine();
}
```

**Comment by Lakasabasz** [comment]
Sorry for late response

My output for your code:
Using lib from nuget
```
64 bit? True
liblsl version: 116
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at SharpLSL.Interop.LSL.lsl_resolve_byprop(IntPtr[], UInt32, System.String, System.String, Int32, Double)
--------------------------------
   at SharpLSL.LSL.Resolve(System.String, System.String, Int32, Int32, Double)
   at Program.<Main>$(System.String[])

Process finished with exit code -1,073,741,819.
```

There is no difference between your build and version from nuget. I modified code adding 2 more Console.WriteLine:
```csharp
using SharpLSL;

Console.WriteLine($"64 bit? {Environment.Is64BitProcess}");

// If this line of code runs successfully, it means
// the liblsl dynamic library has been loaded successfully.
Console.WriteLine($"liblsl version: {LSL.GetLibraryVersion()}");
Console.WriteLine($"liblsl info: {LSL.GetLibraryInfo()}");
Console.WriteLine($"liblsl protocol: {LSL.GetProtocolVersion()}");
```

And output is (for NuGet version):
```
64 bit? True
liblsl version: 116
liblsl info: git:v1.16.2/branch:HEAD/build:/compiler:MSVC-19.41.34120.0/link:SHARED
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at SharpLSL.Interop.LSL.lsl_protocol_version()
--------------------------------
   at SharpLSL.LSL.GetProtocolVersion()
   at Program.<Main>$(System.String[])
```

(For your version):
```
64 bit? True
liblsl version: 116
liblsl info: git:v1.16.2-34-g7e61a2ef/branch:HEAD/build:/compiler:MSVC-19.41.34123.0/link:SHARED
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at SharpLSL.Interop.LSL.lsl_protocol_version()
--------------------------------
   at SharpLSL.LSL.GetProtocolVersion()
   at Program.<Main>$(System.String[])
```

**Comment by Lakasabasz** [comment]
I cloned this repo and I tried to run some examples/tests. Everything that involves calling external method failed. It looks like something at the junction between your code and liblsl (c++ project). Your bindings uses raw build from that project or you add something to it? I'll try to do some std::cout-s in C++ to test where precisely crashes. Current theory is around `SharpLSL.LSL.Resolve` and `var streamInfoPointers = new IntPtr[maxCount];` which provides 1024 null-pointers. Some comments inside C++ project suggests that it should be allocated, but when I did it nothing changed. Can you tell me more about function `lsl_resolve_all`?

**Comment by myd7349** [comment]
Hello! @Lakasabasz Thank you for your reply.

The `liblsl.dll` files in `SharpLSL.Native.win-x64/win-x86` are automatically compiled in https://github.com/myd7349/liblsl-ci-build using GitHub Actions. The corresponding projects in `SharpLSL.Native.win-x64/win-x86` will automatically download the dynamic libraries for the correspond platform and generate NuGet packages. Currently, the liblsl code used for the liblsl-ci-build compilation is version v1.16.2, which you can view here: https://github.com/myd7349/liblsl-ci-build/blob/main/.github/workflows/build.yml#L9. I have not made any modifications to the code, but in the most recent commit of liblsl-ci-build, I added a few configuration options during compilation to generate PDB files along with the Release configuration DLLs for easier debugging. If these configuration changes are the cause of the error, you can use SharpLSL version 1.16.2.1, which does not include those configuration options.

SharpLSL.Native.* 1.16.2.1 uses liblsl-ci-build https://github.com/myd7349/liblsl-ci-build/releases/tag/2024.07.28, and SharpLSL.Native.* 1.16.2.2 uses https://github.com/myd7349/liblsl-ci-build/releases/tag/2024.09.11.

The PInvoke of `lsl_resolve_all` is here: https://github.com/myd7349/SharpLSL/blob/fdce51e263fad63fa46630a2f658bab69d67e397/Source/SharpLSL/Interop/Resolver.Interop.cs#L28

`lsl_resolve_all` is just a C wrapper of original C++ API: https://github.com/sccn/liblsl/blob/v1.16.2/src/lsl_resolver_c.cpp#L50.

**Comment by myd7349** [comment]
By the way, if it's convenient, could you test these versions of liblsl.dll? 

- https://github.com/sccn/liblsl/releases/download/v1.16.2/lsltests-1.16.2-Win_amd64.zip
-  https://github.com/user-attachments/files/17571525/liblsl-7e61a2e-win-x64.zip
- https://github.com/myd7349/liblsl-ci-build/releases/download/2024.07.28/liblsl-v1.16.2-win-x64-dynamic.zip

You just need to replace the liblsl.dll in runtimes/win-x64 with them and then run your executable directly (instead of starting it through the IDE, as that would cause the IDE to copy the SharpLSL.Native.win-x64 dynamic library you referenced into runtimes/win-x64).

**Comment by Lakasabasz** [comment]
I started your executables and interesting effect occurred. `lsltests-1.16.2-Win_amd64.zip` contained in `bin` folder 2 files. `lsl_tests_exported` and `lsl_tests_internal`. In short: exported not working with any dll build, internal works well.

Details of crash (tests not failing, they are crashing):
```
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lsl_test_exported.exe is a Catch v2.13.6 host application.
Run with -? for options

-------------------------------------------------------------------------------
datatransfer - int32_t
-------------------------------------------------------------------------------
D:\a\liblsl\liblsl\testing\ext\DataType.cpp(11)
...............................................................................

D:\a\liblsl\liblsl\testing\ext\DataType.cpp(11): FAILED:
due to a fatal error condition:
  SIGSEGV - Segmentation violation signal

===============================================================================
test cases: 1 | 1 failed
assertions: 1 | 1 failed
```

lslver:
```
LSL version: 116
git:v1.16.2/branch:HEAD/build:/compiler:MSVC-19.40.33812.0/link:SHARED
1222176.953303
```

Exactly same output for git:v1.16.2-34-g7e61a2ef/branch:HEAD/build:/compiler:MSVC-19.41.34123.0/link:SHARED.

But for internal:
```
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lsl_test_internal.exe is a Catch v2.13.6 host application.
Run with -? for options

-------------------------------------------------------------------------------
sleep
-------------------------------------------------------------------------------
D:\a\liblsl\liblsl\testing\int\bench_sleep.cpp(7)
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
sleep1ms                                       100             1     1.56991 s
                                        15.4837 ms    14.8557 ms    15.6516 ms
                                        1.47824 ms    351.036 us    3.43905 ms


2024-11-01 21:05:52.792 (   4.905s) [        41B55F8D]      netinterfaces.cpp:36    INFO| netif '{2E1958B8-2CBB-4D39-AF0B-0990DBE061E8}' (status: 2, multicast: 1
... (logs) ...
===============================================================================
test cases:       28 |       27 passed | 1 failed as expected
assertions: 19408081 | 19408080 passed | 1 failed as expected
```

I try to find exported tests in C++ project and start them with debugger. Probably it is something related to using external dll, and bug is in C++. I'll investigate.

**Comment by Lakasabasz** [comment]
I built from source liblsl.dll and rename it to lsl.dll to fit with expectations and guess: It's working.

I replaced native dll from you with that used before and guess again: It's working using yours samples. What means that bug was inside native dll.

Output from lslver for my build:
```
LSL version: 116
git:v1.16.2-34-g7e61a2ef/branch:HEAD/build:Debug/compiler:GNU-13.1.0/link:SHARED
1224535.989409
```

I should mention that everything I built from native is from submodule inside your repo https://github.com/myd7349/SharpLSL/tree/main/Binding. I suggest you to update submodule, rebuild all native libraries and push version 1.16.3. That fix you mentioned at the beginning possibly was responsible for fixing that issue. Before you releases new version please build new lib and make sure that it contains HEAD from master of liblsl. I don't have MSVC and I am using GCC (MinGW) so I can't check if build from source works.

Still that's very strange you cannot reproduce that error even using different machines

**Comment by myd7349** [comment]
Hello! @Lakasabasz Sorry for the delay.

I have published a new Release:

- SharpLSL: 0.8.0 -> 0.9.0
- SharpLSL.Native: 1.16.2.2 -> 1.16.2.3 (Using the latest commit of liblsl)

And I have created two test projects: [SharpLSLDemo.zip](https://github.com/user-attachments/files/17609726/SharpLSLDemo.zip), with both SendData and ReceiveData.

- SharpLSLDemo-0.0.9-1.16.2.3: uses the latest release of SharpLSL 0.9.0 and SharpLSL.Native 1.16.2.3
-  SharpLSLDemo-0.0.8-1.16.2.1: uses  SharpLSL 0.8.0 and SharpLSL.Native 1.16.2.1

Could you verify these two test projects? Please check if `SharpLSLDemo-0.0.9-1.16.2.3` can run on your computer.

**Comment by Lakasabasz** [comment]
### tl;dr; Please, make sure you build correct commit with native lib, because 1.16.2.2 and 1.16.2.3 are the same.

SharpLSLDemo-0.0.8 (Send and Receive) both:
```
64 bit? True
liblsl version: 116
liblsl library info: git:v1.16.2/branch:HEAD/build:/compiler:MSVC-19.40.33812.0/link:SHARED
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at SharpLSL.Interop.LSL.lsl_resolve_byprop(IntPtr[], UInt32, System.String, System.String, Int32, Double)
--------------------------------
   at SharpLSL.LSL.Resolve(System.String, System.String, Int32, Int32, Double)
   at Program.<Main>$(System.String[])
```

SharpLSLDemo-0.0.9 (Send and Receive) both:
```
64 bit? True
liblsl version: 116
liblsl library info: git:v1.16.2-34-g7e61a2ef/branch:HEAD/build:/compiler:MSVC-19.41.34123.0/link:SHARED
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at SharpLSL.Interop.LSL.lsl_resolve_byprop(IntPtr[], UInt32, System.String, System.String, Int32, Double)
--------------------------------
   at SharpLSL.LSL.Resolve(System.String, System.String, Int32, Int32, Double)
   at Program.<Main>$(System.String[])
```

I checked twice, maybe I started wrong project. But no. `ReceiveData.csproj` contains:
```
﻿<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="SharpLSL" Version="0.0.9" />
    <PackageReference Include="SharpLSL.Native.all" Version="1.16.2.3" />
  </ItemGroup>
</Project>
```

But native dll have strange size. 615KB. Exactly as 1.16.2.2, but csproj uses reference to 1.16.2.3. I checked your repo and liblsl submodule still refers to [6a85d6f](https://github.com/sccn/liblsl/tree/6a85d6f969163cbc17b94b9bd088b972a09a881e). Maybe github didn't update that version? I checked directly on nuget contents of 1.16.2.3 and that was for sure. You pushed wrong content.

**Comment by myd7349** [comment]
Hi! @Lakasabasz Thank you for your tests.

The output of `SharpLSLDemo-0.0.8-1.16.2.1\ReceiveData` should be:

```
64 bit? True
liblsl version: 116
liblsl library info: git:v1.16.2/branch:HEAD/build:/compiler:MSVC-19.40.33812.0/link:SHARED
```

and the out of `SharpLSLDemo-0.0.9-1.16.2.3\ReceiveData` shoud be:

```
64 bit? True
liblsl version: 116
liblsl library info: git:v1.16.2-34-g7e61a2ef/branch:HEAD/build:/compiler:MSVC-19.41.34123.0/link:SHARED
```

and the [last commit of liblsl](https://github.com/sccn/liblsl/commits/master/) is: `7e61a2ef7a25a54b54968fb5c8776054f49b028e`(short version: `7e61a2ef`).

So if you saw `g7e61a2ef`, then the lsl.dll you are using is built from the HEAD of liblsl.

If both projects fail, then the issue definitely lies within liblsl and have not been fixed yet.

> I checked your repo and liblsl submodule still refers to [6a85d6f](https://github.com/sccn/liblsl/tree/6a85d6f969163cbc17b94b9bd088b972a09a881e). Maybe github didn't update that version? I checked directly on nuget contents of 1.16.2.3 and that was for sure. You pushed wrong content.

The submodule here is only used for ClangSharpPInvokeGenerator to analyze header files and generate PInvoke declarations; it will not be used to compile lsl.dll.

**Comment by myd7349** [comment]
I think it would be a good idea to organize the current discussion. I will do this after work.

**Comment by myd7349** [comment]
Things we know(Thank you for your tests):

|                                                              | SharpLSL 0.0.8                                               | SharpLSL 0.0.9                                               |
| ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| lsl.dll downloaded from liblsl v1.16.2 release, https://github.com/sccn/liblsl/releases/download/v1.16.2/lsltests-1.16.2-Win_amd64.zip |                                                              |                                                              |
| SharpLSL.Native.All 1.16.2.1(liblsl v1.16.2)                 | Test 04: [not work](https://github.com/myd7349/SharpLSL/issues/1#issuecomment-2453458000) |                                                              |
| SharpLSL.Native.All 1.16.2.2(liblsl v1.16.2, built with: /Zi /DEBUG /OPT:REF /OPT:ICF) | Test 01: [not work](https://github.com/myd7349/SharpLSL/issues/1#issuecomment-2444869156) |                                                              |
| SharpLSL.Native.All 1.16.2.3(liblsl latest, 7e61a2ef, with patch https://github.com/sccn/liblsl/pull/211, built with: /Zi /DEBUG /OPT:REF /OPT:ICF) | Test 02: [not work](https://github.com/myd7349/SharpLSL/issues/1#issuecomment-2450860602) | Test 05: [not work](https://github.com/myd7349/SharpLSL/issues/1#issuecomment-2453458000) |
| liblsl latest, 7e61a2ef, with patch https://github.com/sccn/liblsl/pull/211, built by yourself | Test 03: [work as expected](https://github.com/myd7349/SharpLSL/issues/1#issuecomment-2452588638) |                                                              |

Currently, the biggest issue is that I am unable to reproduce this problem, either locally or through GitHub Actions: https://github.com/myd7349/SharpLSLDemo .

So, here are the solutions:

1. I can give another try: build liblsl 7e61a2ef without these options: `/Zi /DEBUG /OPT:REF /OPT:ICF`(They are used to generate PDBs in release build) and release a new version of SharpLSL.Native.all 1.16.2.4.
2. If the problem still exists, then replace SharpLSL.Native.all with liblsl built by yourself(just like what you did in Test 03), or download one from https://github.com/sccn/liblsl/releases/tag/v1.16.2 (I do not know if the current release of liblsl v1.16.2 works in your case).
3. If you could provide more information(like what I mentioned here: https://github.com/myd7349/SharpLSL/issues/1#issuecomment-2446471537, https://bentleysystems.service-now.com/community?id=kb_article&sysparm_article=KB0104868), then I can do a mixed mode debugging.
4. If you have Visual Studio installed on your machine, you can also do a mixed mode debugging, just follow this tutorial: https://learn.microsoft.com/en-us/visualstudio/debugger/how-to-debug-managed-and-native-code?view=vs-2022#configure-mixed-mode-debugging .

**Comment by myd7349** [comment]
Hi! @Lakasabasz I have released SharpLSL 0.0.10 and SharpLSL.Native.all 1.16.2.5.
And the test project is also updated: https://github.com/myd7349/SharpLSLDemo.

**Comment by Lakasabasz** [comment]
Thanks for help. I decided to close this issue, because of low impact on others users. That is impossible that only I have this issue and only I can reproduce it. I asked for help my friends and no one was able to crash it.

I built C++ implementation from source repo and replaced my build with that from nuget. It crashed as expected, but differences in build didn't allow me to investigate where exactly was bug. It lies somewhere within resolve stream method. I was able to get asm, but I didn't understand what it means.

There is one thing I am concern. Why your builds are about 610KB (even with optimalization disabled), but my build are ~19MB (Release -O3)? Maybe there is something, but you still won't be able to reproduce it.

To sum up, it is very rare bug that is not possible to recreate. If someone will have the same bug then maybe provide some other details that helps with permanent solution. Thank you for help.

**Comment by myd7349** [comment]
Hello! @Lakasabasz Sorry for the long delay.

Thank you for your effort in tracking down this bug. If you discover anything new, feel free to reopen this issue.

> There is one thing I am concern. Why your builds are about 610KB (even with optimalization disabled), but my build are ~19MB (Release -O3)? Maybe there is something, but you still won't be able to reproduce it.
> 

I built this library using GitHub Actions, and the CMake configuration is as follows:

https://github.com/myd7349/liblsl-ci-build/blob/219edcdda980f85fc4e08ede7d99d0d27d1f0221/.github/workflows/build.yml#L249-L261

Does your local C++ runtime library have any special configurations? You can use the [Dependencies](https://github.com/lucasg/Dependencies) to check the dependencies of `lsl.dll`.

**Comment by myd7349** [comment]
https://github.com/mne-tools/mne-lsl/issues/438

> > OK thanks for the report! My guess is that your systems have the `v142` toolset, corresponding to Visual Studio 2019 and to the required redistributable `Visual C++ 2015-2019`. `liblsl` (https://github.com/sccn/liblsl), from which I assume you downloaded the other `lsl.dll`, is build with that toolset `v142`; while `mne-lsl` did not impose any constraint. That lead `mne-lsl` to build against `v143`, corresponding to Visual Studio 2022 and to the redistributable `Visual C++ 2015-2022 Redistributable` (which I do have on my system).
> > If this is correct, then I'll merge [#440](https://github.com/mne-tools/mne-lsl/pull/440) which pins the build to `v142`, which will have wider compatibility with users. To test it, could you start by:
> > ```
> > * Going to this link: https://github.com/mne-tools/mne-lsl/actions/runs/15649769016?pr=440 (CI run summary of [Pin liblsl build to VS2019 toolset #440](https://github.com/mne-tools/mne-lsl/pull/440))
> > 
> > * Scroll down and download `cibw-wheels-windows-latest-2`
> > 
> > * Extract the content `mne_lsl-1.10.0.dev0-cp310-abi3-win_amd64.whl` from the downloaded `.zip`
> > 
> > * Create an environment, install the wheel and test it:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > uv venv
> > uv pip install mne_lsl-1.10.0.dev0-cp310-abi3-win_amd64.whl
> > uv pip install ipython
> > uv run ipython
> > 
> > [1] import mne_lsl
> > 
> > [2] mne_lsl.lsl.resolve_streams()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > * Report if it worked
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Next, could you also:
> > ```
> > * Uninstall the `Visual C++ 2015-20199 Redistributable`
> > 
> > * Download and install the `Visual C++ 2015-2022 Redistributable` (see: https://learn.microsoft.com/fr-fr/cpp/windows/latest-supported-vc-redist?view=msvc-170#latest-microsoft-visual-c-redistributable-version)
> > 
> > * Test again the PyPI release:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > uv venv
> > uv pip install mne-lsl ipython
> > uv run ipython
> > 
> > [1] import mne_lsl
> > 
> > [2] mne_lsl.lsl.resolve_streams()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Thank you for the report and the help in troubleshooting and resolving this issue!
> 
> Hello! [@mscheltienne](https://github.com/mscheltienne) Your reply reminded me of a similar issue once reported by a user on SharpLSL: [myd7349/SharpLSL#1](https://github.com/myd7349/SharpLSL/issues/1). At the time, I didn't consider the possibility of a MSVC runtime library mismatch—specifically, the runtime library used to build `liblsl` versus the one available on the user's machine.
> 
> Based on your suggestion, I tested it on another PC that only had VS2017 installed (without VS2022), and guess what—I successfully reproduced the issue reported in [myd7349/SharpLSL#1](https://github.com/myd7349/SharpLSL/issues/1).
> 
> Test project: [SharpLSLDemo](https://github.com/myd7349/SharpLSLDemo)
> 
> In the test project:
> 
> <ItemGroup>
>   <PackageReference Include="SharpLSL" Version="0.0.8" />
>   
>   <PackageReference Include="SharpLSL.Native.all" Version="1.16.2.2" />
> </ItemGroup>
> 
> In my testing, I found:
> 
>     * On the PC with only **VS2022** installed (PC A):
>       
>       * `C:\Windows\System32\msvcp140.dll`: 14.42.34433.0
>       * `C:\Windows\System32\vcruntime140.dll`: 14.42.34433.0
>       * `C:\Windows\SysWOW64\msvcp140.dll`: 14.42.34433.0
>       * `C:\Windows\SysWOW64\vcruntime140.dll`: 14.42.34433.0
>       
>       On this machine, **SharpLSLDemo runs normally**.
> 
>     * On the PC with only **VS2017** installed (PC B):
>       
>       * `C:\Windows\System32\msvcp140.dll`: 14.31.31103.0
>       * `C:\Windows\System32\vcruntime140.dll`: 14.31.31103.0
>       * `C:\Windows\SysWOW64\msvcp140.dll`: 14.31.31103.0
>       * `C:\Windows\SysWOW64\vcruntime140.dll`: 14.31.31103.0
>       
>       On this machine, **SharpLSL throws the `System.AccessViolationException`** mentioned in [System.AccessViolationException: Attempted to read or write protected memory myd7349/SharpLSL#1](https://github.com/myd7349/SharpLSL/issues/1).
> 
> 
> I also tested using the `liblsl.dll` built in https://github.com/myd7349/liblsl-ci-build/releases/tag/2025.05.15 on PC B, and it worked without any issues.
> 
> Since `liblsl` is a C++ library, I believe it’s best to **use the MSVC static runtime** when building the DLL. That way, `liblsl` no longer depends on a specific version of `msvcp*.dll` or `vcruntime*.dll`. The `liblsl.dll` from the https://github.com/myd7349/liblsl-ci-build/releases/tag/2025.05.15 release is built this way, using the static MSVC runtime.
